src/coap_session.c

```diff
@@ -1089,15 +1089,25 @@ coap_free_endpoint(coap_endpoint_t *ep) {
   if (ep) {
     coap_session_t *session, *rtmp;
 
-    if (ep->sock.flags != COAP_SOCKET_EMPTY)
-      coap_socket_close(&ep->sock);
-
     SESSIONS_ITER_SAFE(ep->sessions, session, rtmp) {
       assert(session->ref == 0);
       if (session->ref == 0) {
         coap_session_free(session);
       }
     }
+    if (ep->sock.flags != COAP_SOCKET_EMPTY) {
+      /*
+       * ep->sock.endpoint is set in coap_new_endpoint().
+       * ep->sock.session is never set.
+       *
+       * session->sock.session is set for both clients and servers (when a
+       * new session is accepted).
+       *
+       * So, it is safe to call coap_socket_close() after all the sessions
+       * have been freed above as we are only working with the endpoint sock.
+       */
+      coap_socket_close(&ep->sock);
```

Comment:
Maybe we can add some debugging aid here to prevent future breaks:

```c
#ifdef COAP_EPOLL_SUPPORT
  assert(&ep->sock->session == NULL);
#endif /* COAP_EPOLL_SUPPORT */
```

Reply:
Code updated as suggested and pushed.

Reply:
Oops, this should have been

```c
#ifdef COAP_EPOLL_SUPPORT
  assert(ep->sock.session == NULL);
#endif /* COAP_EPOLL_SUPPORT */
```

Reply:
My bad too - must have done the rebuild test with EPOLL disabled - sorry.

Reply:
Test are failing on the assert() - will check out why